### DISCUSSION TITLE
index limits of `Sum` function defined by `parse_expr('...', evaluate=False)` won't evaluate anymore resulting in keyerror

### DISCUSSION THREAD

**Discussion by kompre** [sympy/sympy#26638] [General]
I like to define sympy objects with the `parse_expr` and `evaluate=False` option. This way, combined with the `init_prtinting(order='none'` setup I usually get back an object where the order of various variables appear as I've wrote them. This has worked fine for me for a long time.

Now I want to use the `Sum` function, but the if I use the `evaluate=False` option, then any function which index limit contains an expression that need to be evaluated after substitution, do not get evaluated anymore, even if I try to use the `evalf` and `doit` methods.

Here a minimal working example, using sympy 1.12

```python
from sympy import IndexedBase, Idx, Sum, parse_expr, symbols, ImmutableDenseMatrix as ImMatrix
from sympy import init_printing

# setup the printing environment: in combination with `evaluate=False`, the expression is printed as I've typed it
init_printing(order='none')

a = symbols('a', cls=IndexedBase)
b, n = symbols('b n')
i = symbols('i', cls=Idx)

expr = parse_expr(
    "Sum(a[i]*b, (i, 0, n-1))", # the upper limit is an expression
    local_dict=locals(),
    evaluate=False
)

expr
```
the expression appear like I've intended
![image](https://github.com/sympy/sympy/assets/11796345/479e2f5e-e45e-4b98-83ae-8edd1c13dc39)

Now I substitute the values in
```python
expr.subs(
    {
        a: ImMatrix([0,1,2]),
        b:2,
        n: 3,
    }
)
```
the upper limit appear in the unevalueted form `3-1`:
![image](https://github.com/sympy/sympy/assets/11796345/5c8ebd0c-a3d1-4504-86c3-1bbd1c9433f5)

If I try to evaluate the expression, a key error is raised

```python
expr.subs(
    {
        a: ImMatrix([0,1,2]),
        b:2,
        n: 3,
    }
).evalf() # evaluate it
```
~pay  attention to the line `KeyError: 3`: this means that it is not evaluating the expression `3-1` and discarding the `-1` altogether?~

I've tried changing the value of `n` and I'm still getting back `KeyError: 3`, so appear to not be dependent of the upper limit index value, but connected to the length of the Matrix object.

```py
---------------------------------------------------------------------------
NotImplementedError                       Traceback (most recent call last)
File c:\Users\user\AppData\Local\pypoetry\Cache\virtualenvs\verifica_zavorre-RuYQhVm7-py3.12\Lib\site-packages\sympy\core\evalf.py:1334, in evalf_sum(expr, prec, options)
   1333 if b is not S.Infinity or a is S.NegativeInfinity or a != int(a):
-> 1334     raise NotImplementedError
   1335 # Use fast hypergeometric summation if possible

NotImplementedError: 

During handling of the above exception, another exception occurred:

KeyError                                  Traceback (most recent call last)
File c:\Users\user\AppData\Local\pypoetry\Cache\virtualenvs\verifica_zavorre-RuYQhVm7-py3.12\Lib\site-packages\sympy\polys\matrices\sdm.py:82, in SDM.getitem(self, i, j)
     81 try:
---> 82     return self[i][j]
     83 except KeyError:

KeyError: 3

During handling of the above exception, another exception occurred:

IndexError                                Traceback (most recent call last)
Cell In[28], line 7
      1 expr.subs(
      2     {
...
     89         return self.domain.zero
     90 else:
---> 91     raise IndexError("index out of range")

IndexError: index out of range
```

Changing `evaluate=True` will work, but then the printed expression just looks ugly

**Comment by oscarbenjamin** [comment]
One problem is that `parse_expr` creates an unevaluated `-1*1`:
```python
In [10]: print(expr)
Sum(b*a[i], (i, 0, n - 1*1))
```
It is possible that substituting here should cause that to evaluate:
```
In [11]: print(expr.subs(n, 3))
Sum(b*a[i], (i, 0, 3 - 1*1))
```
If the `-1*1` became `1` then this would simplify to `2`.

Separately the `evalf` code should not raise KeyError in this case and should at least check that the upper limit is not an explicit integer if not handle that case properly.
